### ISSUE TITLE
Jupyter kernel restart during eval after first epoch (MNIST, CPU and CUDA, PyTorch 2.10.0.dev20251114+cu128 on Windows 11 + RTX 5080)

### ISSUE THREAD

**Issue by schlusenbach75** [pytorch/pytorch#167979]
### 🐛 Bug

PyTorch crashes during evaluation after the first training epoch when running a standard MNIST feed-forward network.

There is **no Python traceback**.  
The Jupyter kernel **restarts silently**, indicating a crash in native code (C++ / ATen / MKL / CUDA).

The crash occurs:

- with CUDA **enabled**
- AND with `CUDA_VISIBLE_DEVICES=""` + `device="cpu"`

So the error is not GPU-exclusive.

---

### 🔁 Minimal Reproduction Script

```python
import os
os.environ["CUDA_VISIBLE_DEVICES"] = ""  # hide GPU for this test

import torch
from torch import nn
from torchvision import datasets
from torchvision.transforms import ToTensor
from torch.utils.data import DataLoader

device = "cpu"
print("Device:", device)

train_data = datasets.MNIST("data", train=True, download=True, transform=ToTensor())
test_data  = datasets.MNIST("data", train=False, download=True, transform=ToTensor())

train_loader = DataLoader(train_data, batch_size=80, shuffle=True)
test_loader  = DataLoader(test_data,  batch_size=80, shuffle=True)

class Net(nn.Module):
    def __init__(self):
        super().__init__()
        self.layers = nn.Sequential(
            nn.Flatten(),
            nn.Linear(784,256),
            nn.ReLU(),
            nn.Linear(256,256),
            nn.ReLU(),
            nn.Linear(256,10)
        )
    def forward(self,x):
        return self.layers(x)

model = Net().to(device)
loss_fn = nn.CrossEntropyLoss()
opt = torch.optim.SGD(model.parameters(), lr=0.01)

print("=== TRAIN ===")
for X,y in train_loader:
    X,y = X.to(device), y.to(device)
    pred = model(X)
    loss = loss_fn(pred,y)
    loss.backward()
    opt.step()
    opt.zero_grad()
    break

print("=== EVAL (CRASH HERE) ===")
with torch.no_grad():
    for batch, (X,y) in enumerate(test_loader):
        print("batch", batch)
        X,y = X.to(device), y.to(device)
        pred = model(X)       # → Kernel restart happens here
        break


### Versions

PyTorch version: 2.10.0.dev20251114+cu128
Is debug build: False
CUDA used to build PyTorch: 12.8
OS: Windows 11 Pro (10.0.26200)
Python: 3.12.12 (conda-forge)
GPU: NVIDIA GeForce RTX 5080 (sm_120)
NVIDIA driver: 581.42
torchvision: 0.25.0.dev20251116+cu128
numpy: 2.3.4
MKL: 2025.3.0 (conda-forge)

🔎 Notes

The crash occurs on CPU and CUDA

The crash also occurs in a clean conda environment

Shapes and model are correct, so this is likely a native backend failure

RTX 5080 + CUDA 12.8 + Nightly torch wheels may trigger new ATen failures




cc @seemethere @malfet @atalman @tinglvv @nWEIdia @peterjc123 @mszhanyi @skyline75489 @nbcsm @iremyux @Blackhex

**Comment by albanD** [comment]
Thanks for reporting.
I ran the code locally on my linux machine and everything worked without any issue.
Would you be able to collect a strack trace of the error by any chance?

**Comment by schlusenbach75** [comment]
Thanks for the follow-up. I ran the script again outside Jupyter to collect more detailed information.

Unfortunately, there is no Python stack trace generated by PyTorch itself — because the process is terminated directly by OpenMP. However, I executed the script twice:

Without faulthandler
With faulthandler.enable()
Below are both outputs.

Error without os.environ["KMP_DUPLICATE_LIB_OK"] = "TRUE":

Using cuda device
FeedForwardNetwork(
  (flatten): Flatten(start_dim=1, end_dim=-1)
  (linear_relu_stack): Sequential(
    (0): Linear(in_features=784, out_features=256, bias=True)
    (1): ReLU()
    (2): Linear(in_features=256, out_features=256, bias=True)
    (3): ReLU()
    (4): Linear(in_features=256, out_features=10, bias=True)
  )
)
Epoch 1 (lr: 0.010000)
-------------------------------
loss: 2.314844  [   80/60000]
loss: 2.260088  [ 8080/60000]
loss: 2.187033  [16080/60000]
loss: 2.147110  [24080/60000]
loss: 1.932856  [32080/60000]
loss: 1.633069  [40080/60000]
loss: 1.300159  [48080/60000]
loss: 1.077008  [56080/60000]
Test Error:
 Accuracy: 81.0%, Avg loss: 0.951238

OMP: Error #15: Initializing libomp.dll, but found libiomp5md.dll already initialized.
OMP: Hint This means that multiple copies of the OpenMP runtime have been linked into the program. 
That is dangerous, since it can degrade performance or cause incorrect results. The best thing to 
do is to ensure that only a single OpenMP runtime is linked into the process, e.g. by avoiding static 
linking of the OpenMP runtime in any library. As an unsafe, unsupported, undocumented workaround you 
can set the environment variable KMP_DUPLICATE_LIB_OK=TRUE to allow the program to continue to execute, 
but that may cause crashes or silently produce incorrect results. For more information, please see http://openmp.llvm.org/



After this message the Python process aborts immediately.

---

## 2) Output *with* `faulthandler.enable()`

`faulthandler` does not catch a PyTorch-related stack trace (because OpenMP aborts before a Python exception is raised),  
but it *does* emit a low-level traceback of the interpreter state at the moment of process termination:


FeedForwardNetwork(
  (flatten): Flatten(start_dim=1, end_dim=-1)
  (linear_relu_stack): Sequential(
    (0): Linear(in_features=784, out_features=256, bias=True)
    (1): ReLU()
    (2): Linear(in_features=256, out_features=256, bias=True)
    (3): ReLU()
    (4): Linear(in_features=256, out_features=10, bias=True)
  )
)
Epoch 1 (lr: 0.010000)
-------------------------------
loss: 2.303859  [   80/60000]
loss: 2.266143  [ 8080/60000]
loss: 2.215420  [16080/60000]
loss: 2.143663  [24080/60000]
loss: 1.989288  [32080/60000]
loss: 1.716454  [40080/60000]
loss: 1.376024  [48080/60000]
loss: 1.101562  [56080/60000]
Test Error:
 Accuracy: 79.1%, Avg loss: 0.988641

OMP: Error #15: Initializing libomp.dll, but found libiomp5md.dll already initialized.
OMP: Hint This means that multiple copies of the OpenMP runtime have been linked into the program. 
That is dangerous, since it can degrade performance or cause incorrect results. The best thing to do 
is to ensure that only a single OpenMP runtime is linked into the process, e.g. by avoiding static linking 
of the OpenMP runtime in any library. As an unsafe, unsupported, undocumented workaround you can set the 
environment variable KMP_DUPLICATE_LIB_OK=TRUE to allow the program to continue to execute, but that may 
cause crashes or silently produce incorrect results. For more information, please see http://openmp.llvm.org/
Fatal Python error: Aborted

Thread 0x00006968 (most recent call first):
  <no Python frame>

Current thread 0x000017fc (most recent call first):
  File "C:\Users\Dirk\Work\envs\sandbox\Lib\site-packages\numpy\linalg\_linalg.py", line 669 in inv
  File "C:\Users\Dirk\Work\envs\sandbox\Lib\site-packages\matplotlib\transforms.py", line 1877 in inverted
  File "C:\Users\Dirk\Work\envs\sandbox\Lib\site-packages\matplotlib\transforms.py", line 1463 in __sub__
  File "C:\Users\Dirk\Work\envs\sandbox\Lib\site-packages\matplotlib\axis.py", line 2582 in get_tick_space
  File "C:\Users\Dirk\Work\envs\sandbox\Lib\site-packages\matplotlib\ticker.py", line 2168 in _raw_ticks
  File "C:\Users\Dirk\Work\envs\sandbox\Lib\site-packages\matplotlib\ticker.py", line 2238 in tick_values
  File "C:\Users\Dirk\Work\envs\sandbox\Lib\site-packages\matplotlib\ticker.py", line 2230 in __call__
  File "C:\Users\Dirk\Work\envs\sandbox\Lib\site-packages\matplotlib\axis.py", line 1525 in get_majorticklocs
  File "C:\Users\Dirk\Work\envs\sandbox\Lib\site-packages\matplotlib\axis.py", line 1281 in _update_ticks
  File "C:\Users\Dirk\Work\envs\sandbox\Lib\site-packages\matplotlib\axis.py", line 1351 in get_tightbbox
  File "C:\Users\Dirk\Work\envs\sandbox\Lib\site-packages\matplotlib\artist.py", line 1402 in _get_tightbbox_for_layout_only
  File "C:\Users\Dirk\Work\envs\sandbox\Lib\site-packages\matplotlib\axes\_base.py", line 4564 in get_tightbbox
  File "C:\Users\Dirk\Work\envs\sandbox\Lib\site-packages\matplotlib\artist.py", line 1402 in _get_tightbbox_for_layout_only
  File "C:\Users\Dirk\Work\envs\sandbox\Lib\site-packages\matplotlib\_tight_layout.py", line 82 in _auto_adjust_subplotpars
  File "C:\Users\Dirk\Work\envs\sandbox\Lib\site-packages\matplotlib\_tight_layout.py", line 266 in get_tight_layout_figure
  File "C:\Users\Dirk\Work\envs\sandbox\Lib\site-packages\matplotlib\layout_engine.py", line 188 in execute
  File "C:\Users\Dirk\Work\envs\sandbox\Lib\site-packages\matplotlib\figure.py", line 3640 in tight_layout
  File "C:\Users\Dirk\Work\Data\MNIST_ct_Plot_LR.py", line 230 in plot_loss_and_accuracy
  File "C:\Users\Dirk\Work\Data\MNIST_ct_Plot_LR.py", line 278 in <module>

Extension modules: numpy._core._multiarray_umath, numpy.linalg._umath_linalg, torch._C, torch._C._dynamo.autograd_compiler, 
torch._C._dynamo.eval_frame, torch._C._dynamo.guards, torch._C._dynamo.utils, torch._C._fft, torch._C._linalg, 
torch._C._nested, torch._C._nn, torch._C._sparse, torch._C._special, PIL._imaging, kiwisolver._cext, PIL._imagingft, 
numpy.random._common, numpy.random.bit_generator, numpy.random._bounded_integers, numpy.random._pcg64, numpy.random._mt19937, 
numpy.random._generator, numpy.random._philox, numpy.random._sfc64, numpy.random.mtrand (total: 25)

---

## Interpretation

The abort is clearly triggered by OpenMP before any PyTorch exception or Python-level error can be generated.

This matches the OMP message:

- `libomp.dll` (LLVM OpenMP) → likely from PyTorch nightly pip wheels  
- `libiomp5md.dll` (Intel OpenMP) → pulled in via NumPy + MKL from conda-forge  

Thus two incompatible OpenMP runtimes are loaded simultaneously, resulting in OMP Error #15 and immediate process termination.

---

## Workaround

Adding this at the top of the script:

```python
import os
os.environ["KMP_DUPLICATE_LIB_OK"] = "TRUE"
allows the script to run normally.
With this workaround all 40 epochs complete successfully, plots are generated, and MNIST accuracy reaches ~95%.

Summary
This is an environment-level OpenMP runtime conflict caused by mixing:
pip-installed nightly PyTorch wheels
with conda-forge MKL / NumPy (Intel OpenMP)
Given the above, please feel free to classify or close the issue accordingly.
